python/lsst/cp/pipe/makeBrighterFatterKernel.py

```diff
@@ -0,0 +1,1427 @@
+
+        # stack the individual xcorrs and apply a per-pixel clipped-mean
+        meanXcorr = np.zeros_like(fullCorr)
+        xcorrList = np.transpose(xcorrList)
+        for i in range(np.shape(meanXcorr)[0]):
+            for j in range(np.shape(meanXcorr)[1]):
+                meanXcorr[i, j] = afwMath.makeStatistics(xcorrList[i, j], afwMath.MEANCLIP, sctrl).getValue()
+
+        return self.successiveOverRelax(meanXcorr)
+
+    def successiveOverRelax(self, source, maxIter=None, eLevel=None):
+        """An implementation of the successive over relaxation (SOR) method.
+
+        A numerical method for solving a system of linear equations
+        with faster convergence than the Gauss-Seidel method.
+
+        Parameters:
+        -----------
+        source : `numpy.ndarray`
+            The input array
+        maxIter : `int`, optional
+            Maximum number of iterations to attempt before aborting
+        eLevel : `float`, optional
+            The target error level at which we deem convergence to have occured
+
+        Returns:
+        --------
+        output : `numpy.ndarray`
+            The solution
+        """
+        if not maxIter:
+            maxIter = self.config.maxIterSuccessiveOverRelaxation
+        if not eLevel:
+            eLevel = self.config.eLevelSuccessiveOverRelaxation
+
+        # initialise, and set boundary conditions
+        func = np.zeros([source.shape[0] + 2, source.shape[1] + 2])
+        resid = np.zeros([source.shape[0] + 2, source.shape[1] + 2])
+        rhoSpe = np.cos(np.pi/source.shape[0])  # Here a square grid is assummed
+
+        # Calculate the initial error
+        for i in range(1, func.shape[0] - 1):
+            for j in range(1, func.shape[1] - 1):
+                resid[i, j] = (func[i, j - 1] + func[i, j + 1] + func[i - 1, j] +
+                               func[i + 1, j] - 4*func[i, j] - source[i - 1, j - 1])
+        inError = np.sum(np.abs(resid))
+
+        # Iterate until convergence
+        # We perform two sweeps per cycle,
+        # updating 'odd' and 'even' points separately
+        nIter = 0
+        omega = 1.0
+        dx = 1.0
+        while nIter < maxIter*2:
+            outError = 0
+            if nIter%2 == 0:
+                for i in range(1, func.shape[0] - 1, 2):
+                    for j in range(1, func.shape[0] - 1, 2):
+                        resid[i, j] = float(func[i, j-1] + func[i, j + 1] + func[i - 1, j] +
+                                            func[i + 1, j] - 4.0*func[i, j] - dx*dx*source[i - 1, j - 1])
+                        func[i, j] += omega*resid[i, j]*.25
+                for i in range(2, func.shape[0] - 1, 2):
+                    for j in range(2, func.shape[0] - 1, 2):
+                        resid[i, j] = float(func[i, j - 1] + func[i, j + 1] + func[i - 1, j] +
+                                            func[i + 1, j] - 4.0*func[i, j] - dx*dx*source[i - 1, j - 1])
```

Comment:
Both here and elsewhere in this method you would probably get better speed and possibly code clarity by using `scipy.signal.convolve` to apply one kernel to `func` and one to `source`: `resid = func convolved with [0, 1, 0] [1, 4, 1] [0, 1, 0]] - source convolved with [[0, 0, 0] [0, 0, 0] [dx**2, 0, 0]]`. It's a 1-liner (modulo defining the two kernels, which are always the same and can be created once and re-used).

Reply:
As discussed OOB I'm going to leave this as-is for now because despite being both ugly and slow, it will only save milliseconds of compute to change it, and testing that the behaviour is identical will take a lot longer.